draft-ietf-privacypass-auth-scheme.md

```diff
@@ -454,77 +454,92 @@ token, e.g., the probability that one can forge an authenticator value without
+on avoiding discriminatory behavior across clients when using Privacy Pass tokens.
+
+Clients might also choose to not redeem tokens in subsequent requests when the
+token challenges indicate erroneous or malicious behavior on the part of the
+challenging origin. For example, if a client’s ability to generate tokens via an
+attester and issuer is limited to a certain rate, a malicious origin could send
+an excessive number of token challenges with unique redemption contexts
+in order to cause the client to exhaust its ability to generate new tokens, or
+to overwhelm issuance servers. The limits here will vary based on the specific
+deployment, but clients SHOULD have some implementation-specific policy
+to minimize the number of tokens that can be retrieved by origins.
+
+## Choosing Between Multiple Challenges
+
+A single response from an origin can include multiple token challenges.
+For example, a set of challenges could include different token types
+and issuers, to allow clients to choose a preferred issuer or type.
+
+The choice of which challenge to use for redeeming tokens is up to
+client policy. This can involve which token types are supported or preferred,
+which issuers are supported or preferred, or whether or not the
+client is able to use cached tokens based on the redemption context
+or origin information in the challenge. See {{caching}} for more discussion
+on token caching.
+
+# Origin Behavior {#origin-behavior}
+
+Origins choose what token challenges to send to clients, which will vary
+depending on the use case and deployment model. The origin chooses
+which token types, issuers, redemption contexts, and origin info to include
+in challenges. If an origin sends multiple challenges, each challenge SHOULD
+be equivalent in terms of acceptability for token redemption, since clients
+are free to choose to generate tokens based on any of the challenges.
+
+Origins ought to consider the time involved in token issuance. Particularly,
+a challenge that includes a unique redemption context will prevent a client
+from using cached tokens, and thus can add more latency before the client
+is able to redeem a token.
+
+Origins SHOULD minimize the number of challenges sent to a particular client
+context (referred to as the "redemption context" in
+{{Section 3.3 of ARCHITECTURE}}), to avoid overwhelming clients and issuers
+with token requests that might cause clients to hit rate limits.
+
+## Greasing
+
+In order to prevent clients becoming incompatible with new token challenges,
+origins SHOULD include random token types, from the Reserved list of "greased"
+types (defined in {{token-types}}), with some non-zero probability.
+
+Additionally, for deployments where tokens are not required (such as when tokens
+are used as a way to avoiding showing CAPTCHAs), origins SHOULD randomly 
+choose to not challenge clients for tokens with some non-zero probability.
```

Comment:
I'm really struggling with the incentive structure that surrounds this "SHOULD".  It almost reads like an [OUGHT TO](https://datatracker.ietf.org/doc/html/rfc6919#section-4).

Reply:
```suggestion
Additionally, for deployments where tokens are not required (such as when tokens
are used as a way to avoiding showing CAPTCHAs), origins SHOULD randomly 
choose to not challenge clients for tokens with some non-trivial probability.
This helps origins ensure that their behavior for handling clients that cannot
redeem tokens is maintained and exercised consistently.
```

Modifying the text here to explain. In the CAPTCHA scenario, the CAPTCHA providers already often do some amount of "probing" anyway to keep measuring the background fraud rates, and will choose to show CAPTCHAs instead of getting tokens for some small percentage of the population.